Flag usage of print for python files

vishnu-narayanan opened the pull request:
Linter check to flag usage of print within python files.

Task Link: https://trello.com/c/WgqUxhbQ/6678-t-create-linting-rule-to-chk-on-logging
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Developer Checklist
- [ ] My change requires a change to the documentation.
- [x] I have updated my branch with the latest changes in master.
- [x] I have communicated with the team any changes that might affect work in progress. 
- [ ] I have updated Trello to reflect current state of the feature. 

## Documentation Checklist
- [ ] I have updated the necessary README's
- [ ] I have updated the Runbook

## Description of Changes

- Uses `flake8-print` plugin on top of `flake8` to achieve desired functionality
- Will not block PRs or break devevloper workflows since linter will only run against changed files in PR
- lines/files can be skipped/ignore using standard flake8 ignore rules


## Context
- Improve python codebase by phasing out `print` usage

## Description of Testing Process
- Verified locally
- Ran against template-repo-cf-modules master branch
- Ran against test-branch with change made for python file 
https://internal-commo-jenki-dnef72g28tsg-407778437.us-east-1.elb.amazonaws.com/job/Code-Linter/5392/console
https://github.com/Flux7Labs/template-repo-cf-modules/pull/351

vishnu-narayanan committed:
Add flake8-print to flag print() in python code

vishnu-narayanan committed:
Run pre-commit only for staged files in PR

vishnu-narayanan committed:
Revert "Run pre-commit only for staged files in PR"

This reverts commit 304c451b95d43bc50d940336315d1dcc26a1cc29.

vishnu-narayanan committed:
Add flake8-print to makefile

vishnu-narayanan committed:
Modify linter to check only diff for PRs

This will change the linter behavior to run lint against files changed
in the PR. For nightlies, thsi will run against all the files.

vishnu-narayanan committed:
Change ref in precommit diff check to origin/master

This was failing when the ref just said master. git
rev-parse revealed there was no refs pointing to master
in the cloned directory.

KritikaA06 commented:
Changing Makefile will implement these changes to the rest of the hooks as well, not sure how that goes with other hooks existing. @vishnu-narayanan

shadycuz commented on Makefile:
```diff
@@ -44,13 +44,20 @@ setup:
 	
 	
-	pip install cfn-lint pre-commit flake8 checkov==1.0.580
+	pip install cfn-lint pre-commit flake8 flake8-print checkov==1.0.580
```
Doesn't `additional_dependencies: [flake8-print]` install this for us?

shadycuz replied:
no it doesn't so this should stand and the other should go.

shadycuz commented on .pre-commit-hooks.yaml:
```diff
@@ -44,6 +44,7 @@
   name: flake8-lint
   description: '`flake8` is a command-line utility for enforcing style consistency across Python projects.'
   entry: 'flake8 --ignore=E501,W503'
+  additional_dependencies: [flake8-print]
   language: system
```
This isn't supported for "system". 

https://pre-commit.com/#system
```suggestion
  language: system
```

hanumantharaomvl commented on Makefile:
```diff
@@ -44,13 +44,20 @@ setup:
 	pre-commit clean
-	pre-commit run -c $(HOOK_CONFIG_FILE) --all-files
+
+	if [[ ! -z "$$EVENT_NAME" ]] && [[ "$$EVENT_NAME" == NIGHTLY ]]; then \
+		echo "Running in Nightly Mode"; \
+		pre-commit run -c $(HOOK_CONFIG_FILE) --all-files;\
+	else \
+		pre-commit run -c $(HOOK_CONFIG_FILE) --from-ref origin/master --to-ref HEAD;\
+	fi \
```
@vishnu-narayanan 
1. What if I wanted to re-trigger a build from master with event name MANUAL and wanted to run on all files ?
2. What abour MERGE event ? this also should cover all files.

shadycuz replied:
Why is the CICD job different per event?

If I have a job that deploys a VPC for example. 

Why would I want that job to be different if it's nightly or manual? 

If you have jobs that are different based on the event then it needs to be a new job I think.

vishnu-narayanan replied:
@hanumantharaomvl 
> 1. What if I wanted to re-trigger a build from master with event name MANUAL and wanted to run on all files ?


Yeah, that is a valid question. The alternative is also true right, like if one wants to run MANUAL event with only changes from their branch. At the end of the day, we can decide what we want as a team and implement that way since this is opinionated. 

>2. What abour MERGE event ? this also should cover all files.


Yeah, you are right. I missed this event.

vishnu-narayanan replied:
@shadycuz 

> Why is the CICD job different per event?
If I have a job that deploys a VPC for example.
why would I want that job to be different if it's nightly or manual?
If you have jobs that are different based on the event then it needs to be a new job I think.


Nightly is supposed to run against our master branch to prevent regressions. Manual invocations are an opinionated topic. As I replied in the comment above, the implementation can be what we want as a team.

shadycuz replied:
Maybe for the linter as an expection. 

The problem is people will see this and duplicate it. 

A Jenkins job is a recipe to make X. Not [X,Y,Z]. 

Just keep that in mind for the future that the behavior of a job should be mostly consistent. We had this problem in the past where people tried to build a job that handled all the different use cases and it ended up unreadable and using a dozen large nested if/else statements....

See https://github.com/Flux7Labs/template-repo-jenkins-jobs/blob/991bbb65f7f141b7a620d132de4735b1deb9ab48/jobs/ami-factory/bake-ami/Jenkinsfile

vishnu-narayanan commented:
@hanumantharaomvl @shadycuz 

Right now we have 4 main events from which this job is invoked.
1. Nightly
2. Manual
3. PR Commit
4. PR Merge

Why changing the scope of this job for PR events makes sense ?

1. New linters can be added in future without blocking PRs or breaking developer workflows. This makes sure we will not need to fix everything in every repo before adding a new linter and opens up way to improve codebase incrementally. 
For example, if we merge this new linter without changing the behavior for PR events we will effectively be blocking PRs on all repos.

2. Improved build times for every PR

Can we keep agree to keep the current behavior aka run against all files for events 1, 2, 4 and run only for changed files for 4 ?


cc: @KritikaA06

vishnu-narayanan committed:
Run code-linter for diff when invoke via PR commit event

Run for all files when the job is invoked via Nightly, Manual or
PR Merge events. Run only for diff when invoked via PR commit
event.

vishnu-narayanan committed:
Remove addtional dependencies from flake8

language: system does not supporting dependencies in native precommit way

Co-authored-by: Levi <levi@flux7.com>
